### ISSUE TITLE
 extractor.setVocabulary method does not exist or is not callable in such a way as described in documentation

### ISSUE THREAD

**Issue by ODabrowski** [kyamagu/mexopencv#311]
Hi,

I have installed successfully Matlab R2016A+MEXOpenCV. (Win10 OS).
Please not that I have strictly followed the tutorial https://github.com/kyamagu/mexopencv/wiki/Installation-%28Windows%2C-MATLAB%2C-OpenCV-3%29
and tested both with the recommended 3.1.0 version and the non recommended Master branch version.

I noticed that when trying the basic example given in the documentation:

## Example
 
``` matlab
% create bag of visual words
trainer = cv.BOWKMeansTrainer(K);
dictionary = trainer.cluster(train_descs);

% Compute histogram of visual word occurrences of an image
extractor = cv.BOWImgDescriptorExtractor('SIFT','BruteForce');
extractor.setVocabulary(dictionary);
descs = extractor.compute(im, keypoints);
```

I recieve this error:
```
No appropriate method, property, or field 'setVocabulary' for class 'cv.BOWImgDescriptorExtractor'.

Error in BoW_example1 (line 49)
bowExtractor.setVocabulary(clust);
```

by looking at the corresponding m file it seems that these functions where changed and don't have either the same arguments and are not callable in the same way.
Please could you update the documentation and provide examples on how to use it ?

best regards,

**Comment by ODabrowski** [comment]
it is not possible to call it with *.setVocaculary , but it seems not only this has changed.... The whole library is different

**Comment by amroamroamro** [comment]
That's just a typo in the example, `Vocabulary` is exposed as a property of the object, so you can simply write:

```matlab
extractor = cv.BOWImgDescriptorExtractor('SIFT','BruteForce');
extractor.Vocabulary = dictionary;
```

The library is not really that different, the API in this case was simply adapted to be more MATLAB-friendly... Instead of explicitly calling `getVocabulary` and `setVocabulary`, you can treat it as a regular property.

**Comment by ODabrowski** [comment]
Thanks alot for your amazingly quick answer! I will try that immediately. :)

**Comment by ODabrowski** [comment]
Sorry to bother you again but could you tell me what i'm doing wrong in trying to call it in such a way?:

>> K=100;trainer = cv.BOWKMeansTrainer(K);
>> sift1=cv.SIFT();[k,d]=sift1.detectAndCompute(imread('images/dog1.jpg'));
>> dictionary = trainer.cluster(d);
>> extractor = cv.BOWImgDescriptorExtractor('SIFT','BruteForce');
>> extractor.Vocabulary(dictionary);
Subscript indices must either be real positive integers or logicals.

**Comment by ODabrowski** [comment]
it does not seem to want to take the dictionnary such as it is computed by the trainer

**Comment by ODabrowski** [comment]
sorry for the last post (i deleted it) i made a mistake

**Comment by ODabrowski** [comment]
seems to work fine now: 
I did a little example code like this:
im=imread('images/dog1.jpg');
sift1=cv.SIFT();
[k,d]=sift1.detectAndCompute(im);
K=100;trainer = cv.BOWKMeansTrainer(K);
 dictionary = trainer.cluster(d);

 % Compute histogram of visual word occurrences of an image
 extractor = cv.BOWImgDescriptorExtractor('SIFT','BruteForce');
 extractor.Vocabulary=dictionary;
 descs = extractor.compute(im, k);

**Comment by amroamroamro** [comment]
That looks fine to me. If you want, you can also see the tests we have to show the API in use: https://github.com/kyamagu/mexopencv/blob/master/test/unit_tests/TestBOWImgDescriptorExtractor.m

**Comment by ODabrowski** [comment]
Thanks. yes this is very interesting.

**Comment by ODabrowski** [comment]
Hi again. I hope that i am not bothering you. I would have some questions in order to understand more deeply these libraries. Indeed, I am using this for an academic project and wish not to use them as a "black box".

I have extracted descriptors of 100 images and repeated the above code. I get a dictionnary of 100x128 size. I understand K-Means is used to assign descriptors to clusters. However, could you more precisely explain what this dictionnary contains now in each of these 128 vectors ? (I understand it has something to do with the K-means clusters but now the values in each vector are floating point values and altouhg i gave a K=100 the max val of the dictionnary is 200.1029)

thanks in advance.

**Comment by ODabrowski** [comment]
in fact I would have thought that 1 descriptor would be mapped to 1 cluster in the dict. ?

**Comment by amroamroamro** [comment]
@ODabrowski 
You posted multiple questions in several threads (would have preferred if you asked all in one place), but let me try to give some answers ...

`BOWKMeansTrainer` clusters the descriptors extracted from the training images into K clusters (you choose the number of clusters `K`). Each descriptor is a vector of length D computer from a detected keypoint in an image (each image usually produces multiple keypoints), the vector size depends on the *local feature* algorithm, in this case SIFT gives you a vector of length 128 (i.e. `sift1.descriptorSize()`).

The output dictionary (aka vocabulary or codebook) is a matrix of size K-by-D, one row for each cluster.

The input descriptors given to `BOWKMeansTrainer` is a tall matrix of size (N1+N2+..)-by-K, i.e the concatenation of all descriptors computed from each training image.

**Comment by amroamroamro** [comment]
@ODabrowski 
Now to elaborate (and for my own reference :wink:), the idea of [`cv.BOWImgDescriptorExtractor`](https://kyamagu.github.io/mexopencv/matlab/BOWImgDescriptorExtractor.html) is that it detects and computes descriptors from an image, then uses nearest neighbor (think [`cv.DescriptorMatcher`](https://kyamagu.github.io/mexopencv/matlab/DescriptorMatcher.html)) to match each descriptor to the closest cluster in the vocabulary. This results in a descriptors histogram (usually normalized counts). This histogram is considered an encoding of the image as a feature vector, which is later fed to train the image classifier (like [`cv.SVM`](https://kyamagu.github.io/mexopencv/matlab/SVM.html)).

To learn more about image classification using the [bag-of-words model](https://en.wikipedia.org/wiki/Bag-of-words_model_in_computer_vision), you can see the test included in mexopencv which I mention before:  [test_bow_image_classification](https://github.com/kyamagu/mexopencv/blob/master/test/unit_tests/TestBOWImgDescriptorExtractor.m#L64) in `TestBOWImgDescriptorExtractor`. There's also a similar example I posted on Stack Overflow in the past: http://stackoverflow.com/a/20815886/97160

MathWorks also has some nice explanation with well documented examples (you can still understand it even if you don't use CVST functions but OpenCV ones):
- https://www.mathworks.com/help/vision/ug/image-classification-with-bag-of-visual-words.html
- https://www.mathworks.com/help/vision/examples/image-category-classification-using-bag-of-features.html
